sdk/identity/Azure.Identity/src/IdentityClient.cs

```diff
@@ -0,0 +1,137 @@
+        {
+            Uri authUri = new Uri(_options.AuthorityHost + tenantId + "/oauth2/token");
+
+            using (var request = _pipeline.CreateRequest())
+            {
+                request.SetRequestLine(HttpPipelineMethod.Put, authUri);
+
+                ReadOnlyMemory<byte> content = Serialize(
+                    ( "response_type", "token" ), 
+                    ( "grant_type", "client_credentials" ),
+                    ( "client_id", clientId ),
+                    ( "client_secret", clientSecret ),
+                    ( "scopes", string.Join(" ", scopes) ));
+
+                request.Content = HttpPipelineRequestContent.Create(content);
+
+                var response = await _pipeline.SendRequestAsync(request, cancellationToken).ConfigureAwait(false);
+
+                if (response.Status == 200 || response.Status == 201)
+                {
+                    var result = await DeserializeAsync(response.ContentStream, cancellationToken).ConfigureAwait(false);
+
+                    return new Response<AuthenticationResponse>(response, result);
+                }
+
+                throw await response.CreateRequestFailedExceptionAsync();
+            }
+        }
+
+        private ReadOnlyMemory<byte> Serialize(params (string, string)[] bodyArgs)
+        {
+            var buff = new byte[1024];
+
+            var writer = new FixedSizedBufferWriter(buff);
+
+            var json = new Utf8JsonWriter(writer);
+
+            json.WriteStartObject();
+
+            foreach (var prop in bodyArgs)
+            {
+                json.WriteString(prop.Item1, prop.Item2);
+            }
+
+            json.WriteEndObject();
+
+            return buff.AsMemory(0, (int)json.BytesWritten);
+        }
+
+        private async Task<AuthenticationResponse> DeserializeAsync(Stream content, CancellationToken cancellationToken)
+        {
+            using (JsonDocument json = await JsonDocument.ParseAsync(content, default, cancellationToken).ConfigureAwait(false))
+            {
+                var response = new Dictionary<string, string>();
+
+                string accessToken = null;
+
+                DateTimeOffset expiresOn = DateTimeOffset.MaxValue;
+
+                if (json.RootElement.TryGetProperty("access_token", out JsonElement accessTokenProp))
+                {
+                    accessToken = accessTokenProp.GetString();
+                }
+
+                if(json.RootElement.TryGetProperty("expires_in", out JsonElement expiresInProp))
+                {
+                    expiresOn = DateTime.UtcNow + TimeSpan.FromSeconds(expiresInProp.GetInt64());
+                }
+                
+                return new AuthenticationResponse(accessToken, expiresOn);
+            }
+        }
+
+        // TODO (pri 3): CoreFx will soon have a type like this. We should remove this one then.
+        internal class FixedSizedBufferWriter : IBufferWriter<byte>
```

Comment:
@pakrym, do we have this type in our common sources?

Reply:
We'll have it soon, after updating to new package PR goes in.